nomos-sdp/src/lib.rs

```diff
@@ -0,0 +1,234 @@
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Locator {
+    addr: Multiaddr,
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+pub enum ServiceType {
+    BlendNetwork,
+    DataAvailability,
+    ExecutorNetwork,
+    GenericRestaking,
+}
+
+pub type Nonce = [u8; 16];
+#[derive(Clone, Copy, Eq, PartialEq, Hash, Debug)]
+pub struct ProviderId(pub [u8; 32]);
+#[derive(Clone, Copy, Eq, PartialEq, Hash, Debug)]
+pub struct DeclarationId(pub [u8; 32]);
+#[derive(Clone, Copy, Eq, PartialEq, Hash, Debug)]
+pub struct RewardId(pub [u8; 32]);
+
+#[derive(Clone, Copy, Eq, PartialEq, Hash, Debug)]
+pub struct ProviderInfo {
+    pub provider_id: ProviderId,
+    pub declaration_id: DeclarationId,
+    pub created: BlockNumber,
+    pub rewarded: Option<BlockNumber>,
+    pub withdrawn: Option<BlockNumber>,
+}
+
+impl ProviderInfo {
+    #[must_use]
+    pub const fn new(
+        block_number: BlockNumber,
+        provider_id: ProviderId,
+        declaration_id: DeclarationId,
+    ) -> Self {
+        Self {
+            provider_id,
+            declaration_id,
+            created: block_number,
+            rewarded: None,
+            withdrawn: None,
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Declaration {
+    pub declaration_id: DeclarationId,
+    pub locators: Vec<Locator>,
+    pub services: HashMap<ServiceType, HashSet<ProviderId>>,
+}
+
+impl Declaration {
+    #[must_use]
+    pub fn has_service_provider(&self, service_type: ServiceType, provider_id: ProviderId) -> bool {
+        self.services
+            .get(&service_type)
+            .is_some_and(|service| service.contains(&provider_id))
+    }
+
+    pub fn insert_service_provider(&mut self, provider_id: ProviderId, service_type: ServiceType) {
+        self.services
+            .entry(service_type)
+            .or_default()
+            .insert(provider_id);
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct DeclarationUpdate {
+    pub declaration_id: DeclarationId,
+    pub provider_id: ProviderId,
+    pub service_type: ServiceType,
+    pub locators: Vec<Locator>,
+}
+
+impl<Proof> From<&DeclarationMessage<Proof>> for DeclarationUpdate {
+    fn from(message: &DeclarationMessage<Proof>) -> Self {
+        Self {
+            declaration_id: message.declaration_id(),
+            provider_id: message.provider_id,
+            service_type: message.service_type,
+            locators: message.locators.clone(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct DeclarationMessage<Proof> {
+    pub service_type: ServiceType,
+    pub locators: Vec<Locator>,
+    pub proof_of_funds: Proof,
+    pub provider_id: ProviderId,
+}
+
+impl<Proof> DeclarationMessage<Proof> {
+    fn declaration_id(&self) -> DeclarationId {
+        let mut hasher = Blake2b::new();
+        for locator in &self.locators {
+            hasher.update(locator.addr.as_ref());
+        }
+        DeclarationId(hasher.finalize().into())
+    }
+}
```

Comment:
@madxor The spec doesn't specify the hasher. We use blake2b but I think we should specify it in the spec as well. As it is not really implementation details. Someone that has to implement the same should use the same hasher.